### ISSUE TITLE
Aikau Unit Test page does not load correctly on Windows

### ISSUE THREAD

**Issue by AFaust** [Alfresco/Aikau#1190]
Due to my problems running the test suite via Grunt (#1185) I tried to manually access the unit test page(s). In my Windows 7 environment, the Jetty server fails to properly process the dependencies of the test page and as a result the unit test page does not load correctly. Hundreds/thousands of error messages are logged in the following form:

2016-08-03 14:44:08.499:WARN:oejur.FileResource:qtp53688719-26: bad alias for F:\Aikau\aikau\target\test-classes\testApp\classpath*:js\aikau\1.0.80-SNAPSHOT\alfresco\menus\css\images\transparent-20-debug.png: java.io.IOException: The filename, directory name, or volume label syntax is incorrect

Running the same branch in my Linux VM (Mint 17.3), I can access the unit test page without such issues. I have tried both JDK 7 and JDK 8 on Windows and neither make a difference (I only have JDK 7 on the VM). The primary difference between both environments is that the VM actually has Node installed (as part of #1185) while the Windows environment has not - so could this be a side-effect of Surf now trying to use Node for LESS processing / aggregation + compression?

Tests where done on develop branch of 1.0.79-SNAPSHOT / 1.0.80-SNAPSHOT states.

**Comment by draperd** [comment]
I have a Windows 10 VM that I can try and get setup to run everything on and look into this. There could be an issue with the paths required by LESS. I'll try to find the time to go through the setup on Windows and see if I can reproduce the problems, but I'm not sure how quickly I'll be able to get to that.

**Comment by draperd** [comment]
Quick question... are you using Cygwin? Someone pointed out to me the path separator in the error (e.g “:” instead of “;”)

**Comment by AFaust** [comment]
No - Cygwin is on my ban-list, or to be more precise I don't accept any software that only supports Windows when run within Cygwin.

**Comment by draperd** [comment]
OK... it looks like the issue is because of changes we made with regards to the LESS processing since the instructions were originally written. In Share (by default) we're still using a Java based LESS compiler, but it is possible to use an external Node based compiler as it's much, much faster.

If you look in the [/aikau/src/test/resources/testApp/WEB-INF/config/web-application-config.xml](https://github.com/Alfresco/Aikau/blob/0f27bb5eb45dd16be4d08193a7a6f417db96d0f1/aikau/src/test/resources/testApp/WEB-INF/config/web-application-config.xml) file you'll find the bean override for the `css.theme.handler`.

It looks like this:

```XML
<!-- Use an external Node driven LESS CSS theme handler -->
<bean id="css.theme.handler" parent="css.theme.handler.abstract" class="org.springframework.extensions.surf.ExternalLessCssThemeHandler">
   <!-- Node LESS compiler command - the hyphen "-" in the cmd value means 'read from stdin'
        This default works well on Linux/Mac with NPM installed. On Windows, a full path will probably be required, for example:
        "C:\Users\rsmith2\AppData\Roaming\npm\lessc.cmd -" -->
   <property name="cmd"><value>${basedir}/node_modules/less/bin/lessc -</value></property>
</bean>
```

Note that we even spotted that there would be Windows issues, but we failed to update the setup instructions... apologies for that.

Anyway, if you change the "cmd" property to use the correct executable then it should work. So for example, following the setup instructions this would be

```
C:\cygwin64\devel\Aikau\aikau\node_modules\.bin\lessc.cmd
```

...this would be the Cygwin based path obviously (which you're not using). However, you should be able to find the appropriate location of your installed LESS executable. If you install LESS globally, e.g.

```
npm install -g less
```

...then you can expect to find the executable in a location similar to the location in the comment (e.g. user specific). Hopefully this should resolve your problem. Please let me know if it doesn't!

**Comment by draperd** [comment]
I've updated the [Windows setup instructions](https://github.com/Alfresco/Aikau/wiki/Aikau-Setup-(Windows))

**Comment by AFaust** [comment]
It would really be cool if Aikau / Surf / Share had a central config file like alfresco-global.properties for configuration changes like these, so people don't have to duplicate + adapt XML configuration files later in production (I also don't like having to mess with source-controlled test files to make something work that is environment specific). Similar to the path, users might want to easily switch between JVM LESS and Node LESS.
I know this is actually Surf related - so basically only a strong opinion to pass on and hopefully ponder.

**Comment by draperd** [comment]
I don't disagree with that.

The problem as always one of resourcing and requirement. At the moment it would seem that you're the only person outside of Alfresco that has attempted to get the development environment up and running on Windows or even on a Linux VM. 

It's something of a chicken and egg situation. If we invest in this kind of infrastructure changes, will it guarantee that anyone actually makes use of them? Or do we wait to get problems reported (such as yours) that we can then address.

With limited resources working on Aikau / Surf it's important that we try to get the right balance and work on the right things to move it forwards. I'm hoping that you'll be able to lead by example, getting the unit tests running and start providing full pull requests that you've already been able to fully regression test.

If we can get more people contributing to Aikau, and see that their contributions are being accepted then that might encourage further participation to improve infrastructure as well as the code.

**Comment by draperd** [comment]
Did the change resolve the problem for you though? Can you get the test app running? And are you happy to close this issue?

**Comment by AFaust** [comment]
I had to re-install Node which I uninstalled as fast as I could when switching my attempts to the Linux VM. Even if given a proper Windows path to lessc, the same errors (WARN statements) are logged and the page does not load correctly.

The IOExceptions / WARN statements occur when the DependencyHandler is executing ```getResourceInputStream(String path, DependencyHandlerProcessingCallback callback)```. The code to resolve a resource via the Spring application context uses an incorrect protocol prefix. The JavaDoc of ```getResource(String)``` clearly states that it supports the ```classpath:``` prefix - the calling code uses ```classpath*:``` which can't be handled and results in fallback code. It is the WinNTFileSystem that triggers the IOException during path canonicalization. It is very likely the other Java implementations of a filesystem (on Linux or Mac) can deal with the effective path in the like of ```F:\Aikau\aikau\target\test-classes\testApp\classpath*:js\aikau\1.0.80-SNAPSHOT\alfresco\core\ProcessWidgets.js```. So this is a **bug** in the Surf DependencyHandler.

The LESS error I currently get (unfortunately not logged - have to use debugger + watch expressions to manually read from stdErr) is "lessc: no input files" when the external less processor appends the less variables to stdIn.
